test/libs/data-structures/SparseMerkleTree.test.ts

```diff
@@ -0,0 +1,343 @@
+          PoseidonUnit3L: await (await getPoseidon(3)).getAddress(),
+        },
+      });
+      const newMerkleTree = await SparseMerkleTreeMock.deploy();
+
+      await expect(newMerkleTree.getUintRoot()).to.be.rejectedWith("SparseMerkleTree: tree is not initialized");
+      await expect(newMerkleTree.addUint(1n, 1n)).to.be.rejectedWith("SparseMerkleTree: tree is not initialized");
+    });
+
+    it("should build a Merkle Tree of a predefined size with correct initial values", async () => {
+      const value = 2341n;
+      const index = BigInt(poseidonHash(ethers.toBeHex(value)));
+
+      await merkleTree.addUint(index, value);
+
+      await localMerkleTree.add(index, value);
+
+      expect(await merkleTree.getUintRoot()).to.equal(await getRoot(localMerkleTree));
+
+      expect(await merkleTree.getUintMaxDepth()).to.equal(20);
+      expect(await merkleTree.getUintNodesCount()).to.equal(1);
+
+      await compareNodes(await merkleTree.getUintNode(1), index);
+      await compareNodes(await merkleTree.getUintNodeByIndex(index), index);
+
+      const onchainProof = getOnchainProof(await merkleTree.getUintProof(index));
+
+      expect(await verifyProof(await localMerkleTree.root(), onchainProof, index, value)).to.be.true;
+    });
+
+    it("should build a Merkle Tree correctly with multiple elements", async () => {
+      for (let i = 1n; i < 20n; i++) {
+        const value = BigInt(ethers.toBeHex(ethers.hexlify(ethers.randomBytes(28)), 32));
+        const index = BigInt(poseidonHash(ethers.toBeHex(`0x` + value.toString(16), 32)));
+
+        await merkleTree.addUint(index, value);
+
+        await localMerkleTree.add(index, value);
+
+        expect(await merkleTree.getUintRoot()).to.equal(await getRoot(localMerkleTree));
+
+        await compareNodes(await merkleTree.getUintNodeByIndex(index), index);
+
+        const onchainProof = getOnchainProof(await merkleTree.getUintProof(index));
+        expect(await verifyProof(await localMerkleTree.root(), onchainProof, index, value)).to.be.true;
+      }
+
+      expect(await merkleTree.isUintCustomHasherSet()).to.be.true;
+
+      await expect(merkleTree.setUintPoseidonHasher()).to.be.rejectedWith("SparseMerkleTree: tree is not empty");
+    });
+
+    it("should generate empty proof on empty tree", async () => {
+      const onchainProof = getOnchainProof(await merkleTree.getUintProof(1n));
+
+      expect(onchainProof.allSiblings()).to.have.length(0);
+    });
+
+    it("should generate an empty proof for but with aux fields", async () => {
+      await merkleTree.addUint(7n, 1n);
+
+      const onchainProof = await merkleTree.getUintProof(5n);
+
+      expect(onchainProof.auxIndex).to.equal(7n);
+      expect(onchainProof.auxValue).to.equal(1n);
+      expect(onchainProof.auxExistence).to.equal(true);
+      expect(onchainProof.existence).to.equal(false);
+    });
+
+    it("should reset the value of the node", async () => {
```

Comment:
Does original SMT do that?

Reply:
Yep, it does with a lot of asserts:

1. Initial call: https://github.com/iden3/contracts/blob/8815bf53e989311d94301f4c513436c1ff776911/contracts/lib/SmtLib.sol#L463
2. Asserts: https://github.com/iden3/contracts/blob/8815bf53e989311d94301f4c513436c1ff776911/contracts/lib/SmtLib.sol#L553